### ISSUE TITLE
PrimTorch decomps for random functions

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#77659]
I'd like to see primtorch decompositions that map all randomness to a single function (perhaps aten.rand()).

The snag: I want the same answer as eager mode.  Same as eager mode means I can check correctness.  It is also a better experience for users.

The ones I am seeing in torchbench models are:
- [ ] dropout (already exists in torch._decomp, but different(?) answer than eager)
- [ ] randn
- [ ] randperm

cc @pbelevich @ezyang @mruberry @ngimel @Lezcano @peterbell10 @Chillee

**Comment by ngimel** [comment]
Given that triton random number generators https://triton-lang.org/master/python-api/triton.language.html#random-number-generation will necessarily produce different numbers than the one we use in eager, and that triton kernels will traverse input differently than how it's traversed in eager (and thus even if the stream of random numbers was the same it would be mapped to locations differently) getting the same answer as eager mode is very challenging and also severely limiting available optimization strategies. 
That said, having decompositions for random functions is reasonable. 
Dropout in torch._decomp does give different answer than eager, and like with reductions, making sure that the 2 match is pretty hard.

**Comment by jansel** [comment]
Yeah, I am aware same-as-eager is hard. I think the *decomps* should be same-as-eager, though possibly not the core random number generator.

1) This allows easier correctness checking: we can swap in the eager version of `rand()` when correctness checking, then use the faster `rand()` when benchmarking.
2) If at some point in the future we do want to make the backends give same-as-eager results, we only need to change the implementation of one function.

Basically I want to contain differences from eager to a few places in the system as possible.

**Comment by gchanan** [comment]
We don't give the same randomness across CPU and CUDA do we?  If not, how could the decomps match eager?

**Comment by ezyang** [comment]
I'm guessing we're going to have to expose some more, Philox-oriented primitives around explicitly manipulating the RNG stream offset to actually manage this.

**Comment by ezyang** [comment]
@gchanan let's make life easier for us, and say we just want the CUDA behavior to match :>

**Comment by jansel** [comment]
We could also change the behavior of eager to make it easier to replicate.

**Comment by mruberry** [comment]
https://github.com/pytorch/pytorch/pull/78026 adds a Python reference for nn.functional.dropout, but it's not bitwise-accurate with either the eager CPU or CUDA implementations, which are distinct from each other. Modeling the way random bits are used by either strategy will require additional work, and if we want the same dropout pattern for the CPU and CUDA we might have to change eager mode, too. It's an interesting question whether we should require trace executors implement the same dropout patterns (or more generally, the same random generation).

**Comment by ezyang** [comment]
This program demonstrates that it is in principle possible to line up uniform and dropout:

```
import torch

torch.cuda.manual_seed(0xDEADBEEF)

b_old = torch.cuda.get_rng_state()

x = torch.empty(256, device='cuda')
x.uniform_()

b_new = torch.cuda.get_rng_state()

torch.cuda.set_rng_state(b_old)
i = torch.ones(256, device='cuda')
#r = torch.nn.functional.dropout(i)
r, _ = torch._C._VariableFunctions.native_dropout(i, 0.5, True)

b_new2 = torch.cuda.get_rng_state()

print(r[::4])
print(torch.where(x <= 0.5, i, 0))
```

This prints:

```
tensor([0., 0., 0., 0., 0., 2., 0., 2., 0., 0., 0., 2., 2., 2., 0., 2., 0., 0.,
        0., 2., 2., 0., 2., 0., 2., 0., 0., 2., 2., 2., 2., 2., 0., 2., 2., 2.,
        2., 2., 0., 0., 0., 0., 2., 2., 0., 2., 0., 0., 2., 2., 0., 2., 0., 2.,
        2., 2., 2., 2., 2., 2., 2., 0., 0., 0.], device='cuda:0')
tensor([0., 0., 0., 0., 0., 1., 0., 1., 0., 0., 0., 1., 1., 1., 0., 1., 0., 0.,
        0., 1., 1., 0., 1., 0., 1., 0., 0., 1., 1., 1., 1., 1., 0., 1., 1., 1.,
        1., 1., 0., 0., 0., 0., 1., 1., 0., 1., 0., 0., 1., 1., 0., 1., 0., 1.,
        1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 1., 0.,
        1., 1., 0., 0., 0., 1., 1., 1., 1., 1., 0., 0., 1., 1., 0., 1., 0., 0.,
        0., 1., 1., 0., 1., 1., 0., 0., 0., 0., 0., 1., 0., 1., 1., 0., 0., 0.,
        0., 0., 1., 1., 0., 1., 1., 0., 0., 1., 1., 1., 1., 0., 0., 0., 1., 1.,
        0., 1., 0., 1., 0., 1., 0., 1., 1., 0., 1., 1., 1., 1., 1., 1., 1., 1.,
        0., 1., 0., 0., 0., 0., 1., 0., 1., 1., 1., 1., 0., 0., 0., 0., 1., 0.,
        0., 1., 0., 0., 1., 1., 1., 0., 1., 1., 1., 1., 0., 0., 1., 1., 1., 1.,
        1., 0., 1., 1., 1., 0., 0., 0., 1., 0., 1., 1., 0., 1., 0., 0., 1., 1.,
        1., 0., 0., 0., 0., 0., 0., 1., 1., 0., 1., 1., 1., 0., 1., 0., 1., 1.,
        0., 1., 1., 0., 0., 0., 0., 1., 0., 0., 1., 0., 1., 0., 0., 0., 0., 0.,
        0., 0., 1., 1., 0., 1., 1., 0., 1., 0., 0., 0., 0., 0., 0., 0., 1., 0.,
        0., 1., 0., 0.], device='cuda:0')
```

The step size 4 is required because our uniform implementation uses `curand_normal4` inefficiently and drops three out of four samples. This will need to be fixed.

**Comment by ngimel** [comment]
It's possible in principle for contiguous aligned tensors (with fixes to the eager distribution, as you point out). Worse alignment or non-contiguous will make it harder.

**Comment by ngimel** [comment]
Hm, also that's weird, I checked the impl, and uniform_ should be using all 4 numbers that it generated, it's just putting them to different places in the tensor, and it'll become painfully obvious once we start running and comparing results on larger tensors.

**Comment by ngimel** [comment]
This line https://github.com/pytorch/pytorch/blob/cac16e2ee293964033dffa6616f78b68603cd565/aten/src/ATen/native/cuda/DistributionTemplates.h#L86 will use all `unroll_factor` generated numbers, and `unroll_factor` is 4, I checked.

**Comment by ezyang** [comment]
Oh, pointer arithmetic on struct fields, 'das some weird shit

**Comment by ngimel** [comment]
My fault, but what would you suggest for `.x`, `.y`, `.z`, `.w` fields instead without manually unrolling the loop?

**Comment by ezyang** [comment]
No other good choices :) I guess probably good to have a comment haha

**Comment by Lezcano** [comment]
A side note: 
From an implementation perspective, I don't know how it's done in CUDA, but on CPU, say on the STL or boost, the algorithms to generate, say, a number distributed as a normal 0,1 are **highly** non-trivial (moved away from the Box Muller algorithm to the Ziggurat algorithm years ago, see [[here](https://github.com/boostorg/random/blob/develop/include/boost/random/normal_distribution.hpp)]).
Modelling all the random operations from one random source seems **very** technically challenging, if you want efficient algorithms...

**Comment by ezyang** [comment]
@zou3519, @albanD and I knocked heads on this some more today. The non-contiguous access pattern is difficult be we think it is also reverse engineerable. More at

https://docs.google.com/document/d/1xuBGdEppsdMZB4_ygQJ-sb6Am6Nh296dlZ3p8y6CcoQ/edit#

cc @anijain2305
